src/plugins/data/common/search/search_source/query_to_fields.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { DataViewLazy } from '@kbn/data-views-plugin/common';
+import { fromKueryExpression, getKqlFieldNames } from '@kbn/es-query';
+import type { SearchRequest } from './fetch';
+import { EsQuerySortValue } from '../..';
+
+export async function queryToFields({
+  dataView,
+  sort,
+  request,
+}: {
+  dataView: DataViewLazy;
+  sort?: EsQuerySortValue | EsQuerySortValue[];
+  request: SearchRequest;
+}) {
+  let fields = dataView.timeFieldName ? [dataView.timeFieldName] : [];
+  if (sort) {
+    const sortArr = Array.isArray(sort) ? sort : [sort];
+    fields.push(...sortArr.flatMap((s) => Object.keys(s)));
+  }
+  for (const query of request.query) {
+    if (query.query && query.language === 'kuery') {
+      const nodes = fromKueryExpression(query.query);
+      const queryFields = getKqlFieldNames(nodes);
+      fields = fields.concat(queryFields);
+    }
+  }
```

Comment:
nit: if the `let` usage is only due to the `concat` call at the end, maybe that can be replaced with `push(...queryFields)` and make this a `const`

```suggestion
  const fields = dataView.timeFieldName ? [dataView.timeFieldName] : [];
  if (sort) {
    const sortArr = Array.isArray(sort) ? sort : [sort];
    fields.push(...sortArr.flatMap((s) => Object.keys(s)));
  }
  for (const query of request.query) {
    if (query.query && query.language === 'kuery') {
      const nodes = fromKueryExpression(query.query);
      const queryFields = getKqlFieldNames(nodes);
      fields.push(...queryFields);
    }
  }
```

Reply:
Although this nit makes sense, I prefer to leave this PR as close to its original counterpart (https://github.com/elastic/kibana/pull/184890) as possible